### PR TITLE
DE46041: Fixed the metaspace leak when managing/evaluating policies

### PR DESCRIPTION
**Pull request by anubhavi25**

### CODE REVIEW

**Comment by anubhavi25** [comment]
Includes the following changes:
- [x] Added the Grengine library (https://www.grengine.ch/manual.html) that wraps Groovy with much better garbage collection and smarter script caching/reloading
- [x] Moved Groovy-script-related logic from `GroovyConditionScript` to `GroovyConditionShell` (since execution of compiled scripts are better handled by the Grengine shell)
- [x] Removed the `@Component` annotation on `GroovyConditionShell` to prevent the Spring IoC container from holding onto the last reference even when removed from `GroovyConditionCache` (and thus leak metaspace)
- [x] Cached condition scripts and the shells they execute in are now removed when a policy set is deleted (which in turn also unloads all related classes from the relevant classloaders)
- [x] Enhanced the top-level `start-acs.sh` script to allow passing custom JVM options (i.e. decreasing the max metaspace size to deliberately create memory pressure while integration testing)
- [x] Moved tests in `GroovyConditionScriptTest` to `GroovyConditionShellTest` and updated other unit tests as necessary

**Comment by anubhavi25** [comment]
Passing integration pull request builds:

https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/998/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/999/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1000/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1001/

**Comment by anubhavi25** [comment]
Includes the following changes:
- [x] A singleton `GroovyConditionShell` is `@Autowired` everywhere instead of creating multiple instances 
- [x] Cached condition scripts are now removed when a policy set is deleted
- [x] Enhanced the top-level `start-acs.sh` script to allow passing custom JVM options (i.e. decreasing the max metaspace size to deliberately create memory pressure while integration testing)
- [x] Updated unit tests as necessary

Related configuration and testing tasks:

- [x] New Relic monitoring of the CF release candidate app under load from its synthetic monitor with added memory pressure
- [x] Increase metaspace memory for all environments if necessary
- [x] Run performance tests against this branch (see story description for more details)

**Comment by anubhavi25** [comment]
Passing integration pull request builds:

https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1002/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1003/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1004/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1006/

File: service/src/main/java/com/ge/predix/acs/service/policy/admin/PolicyManagementServiceImpl.java [hunk 1]
```diff
@@ -136,6 +136,12 @@ public void deletePolicySet(final String policySetId) {
         if (policySetEntity != null) {
             LOGGER.info("Found an existing policy set policySetName={}, zone={}, deleting now.", policySetId,
                     zone.getName());
+
+            PolicySet policySet = this.jsonUtils.deserialize(policySetEntity.getPolicySetJson(), PolicySet.class);
+            if (policySet != null) {
+                this.policySetValidator.removeCachedConditions(policySet);
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/service/policy/admin/PolicyManagementServiceImpl.java:142]: @anubhavi25, cached conditions are shared between policy sets.  Cached conditions that are removed when policy set is deleted might be also used by some other remaining policy sets. 

Policy Set One:
Policy One: condition 1, condition 2

Policy Set Two:
Policy Two: condition 1 

Removing cached "condition 1" when "Policy Set One" is deleted will remove it for "Policy Set Two" as well.

Seems like we are trading off the performance for memory usage as conditions will have to be recompiled again when these remaining policy sets are evaluated.  If degradation is noted during performance test we might need to reconsider that change.

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java [hunk 2]
```diff
@@ -231,9 +228,6 @@ private PolicyEvaluationResult evalPolicySet(final PolicySet policySet, final St
 
                 boolean conditionEvaluationResult = true;
                 if (!policy.getConditions().isEmpty()) {
-                    if (null == groovyShell) {
-                        groovyShell = new GroovyConditionShell();
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationServiceImpl.java:235]: 👍  Good catch.  Allocated shell is never used.

**Review by irinaepshteyn — APPROVED** [review]

File: service/src/main/java/com/ge/predix/acs/service/policy/admin/PolicyManagementServiceImpl.java (hunk 1 above)
Comment (anubhavi25) [reply service/src/main/java/com/ge/predix/acs/service/policy/admin/PolicyManagementServiceImpl.java:142]: @irinaepshteyn performance tests results didn't show a noticeable impact to latency (see story description for details)

**Pull request merged**